### DISCUSSION TITLE
CRUD sqlite sample in textual textualize

### DISCUSSION THREAD

**Discussion by bisoogo** [Textualize/textual#2082] [Q&A]
Hello
I am very much excited with Textual precisely what I've been looking for text-based RAD. Might there be sample crud code based on sqlite3? Thank you much

**Comment by bisoogo** [comment]
So suppose I have a Python crud script, how do I integrate it into Textual?

**Reply by Hisham-ghanayem** [reply]
Answre by AI :) 
Integrating an existing Python CRUD script into [Textual](https://textual.textualize.io/) involves wrapping your logic within a subclass of the App class and mapping your CRUD functions to interactive widgets. 
Real Python
Real Python
 +1
1. Structure Your App
Subclass App and use the compose() method to yield the widgets that will form your interface. 
Real Python
Real Python
 +1
Read: Use a [DataTable](https://textual.textualize.io/widgets/data_table/) widget to display your records.
Create/Update: Use Input fields and Buttons to collect user data and trigger your script’s logic.
Delete: Bind keys (e.g., d for delete) or use a button to remove the selected row. 
Hashnode
Hashnode
 +4
2. Map CRUD Logic to Events
Connect your existing script's functions to Textual's event handlers: 
on_mount: Call your "Read" function here to populate the DataTable when the app starts.
on_button_pressed: Trigger "Create" or "Update" functions when a user clicks a "Submit" or "Save" button.
Actions & Bindings: Map keyboard shortcuts to specific CRUD actions using the BINDINGS list. 
Hashnode
Hashnode
 +4
3. Example Integration Pattern
You can follow this general structure to bridge your script with the TUI:
python
from textual.app import App, ComposeResult
from textual.widgets import DataTable, Input, Button
# Import your existing CRUD functions
from my_crud_script import get_all_data, add_record

class CRUDApp(App):
    def compose(self) -> ComposeResult:
        yield Input(placeholder="Enter name", id="name_input")
        yield Button("Add Record", variant="success")
        yield DataTable()

    def on_mount(self) -> None:
        table = self.query_one(DataTable)
        table.add_columns("ID", "Name")
        # Read from your script
        data = get_all_data() 
        table.add_rows(data)

    def on_button_pressed(self, event: Button.Pressed) -> None:
        name = self.query_one("#name_input").value
        # Create using your script
        add_record(name) 
        # Refresh the UI table here...

if __name__ == "__main__":
    CRUDApp().run()

**Reply by willmcgugan** [reply]
Please don't waste people's time with unhelpful slop.
